examples/simple-chatbot/client/react-native/src/context/VoiceClientContext.tsx

```diff
@@ -0,0 +1,229 @@
+  timerCountDown: number
+  // methods
+  start: (url: string) => Promise<void>
+  leave: () => void
+  toggleMicInput: () => void
+  toggleCamInput: () => void
+}
+
+export const VoiceClientContext = createContext<VoiceClientContextProps | undefined>(undefined)
+
+interface VoiceClientProviderProps {
+  children: ReactNode
+}
+
+export const VoiceClientProvider: React.FC<VoiceClientProviderProps> = ({ children }) => {
+  const [voiceClient, setVoiceClient] = useState<RTVIClient | null>(null)
+  const [inCall, setInCall] = useState<boolean>(false)
+  const [currentState, setCurrentState] = useState<TransportState>("disconnected")
+  const [botReady, setBotReady] = useState<boolean>(false)
+  const [isMicEnabled, setIsMicEnabled] = useState<boolean>(false)
+  const [isCamEnabled, setIsCamEnabled] = useState<boolean>(false)
+  const [videoTrack, setVideoTrack] = useState<MediaStreamTrack>()
+  const [localAudioLevel, setLocalAudioLevel] = useState<number>(0)
+  const [remoteAudioLevel, setRemoteAudioLevel] = useState<number>(0)
+  const [timerCountDown, setTimerCountDown] = useState<number>(0)
+
+  const botSpeakingRef = useRef(false)
+  let meetingTimer: NodeJS.Timeout | null
+
+  const createVoiceClient = useCallback((url: string): RTVIClient => {
+    return new RTVIClient({
+      transport: new RNDailyTransport(),
+      params: {
+        baseUrl: url,
+        endpoints: {
+          connect: "/connect"
+        }
+      },
+      enableMic: true,
+      enableCam: false
+    })
+  }, [])
+
+  const handleError = useCallback((error: any) => {
+    console.log("Error occurred:", error)
+    const errorMessage = error.message || error.data?.error || "An unexpected error occurred"
+    Toast.show({
+      type: 'error',
+      text1: errorMessage,
+    })
+  }, [])
+
+  const setupListeners = useCallback((voiceClient: RTVIClient): void => {
+    const inCallStates = new Set(["authenticating", "connecting", "connected", "ready"])
+
+    voiceClient
+      .on("transportStateChanged", (state: TransportState) => {
+        setCurrentState(voiceClient.state)
+        setInCall(inCallStates.has(state))
+      })
+      .on("error", (error: RTVIMessage) => {
+        handleError(error)
+      })
+      .on("botReady", () => {
+        setBotReady(true)
+        let expirationTime = voiceClient.transportExpiry
+        if (expirationTime) {
+          startTimer(expirationTime)
+        }
+      })
+      .on("disconnected", () => {
+        setBotReady(false)
+        stopTimer()
+        setIsMicEnabled(false)
+        setIsCamEnabled(false)
+      })
+      .on("localAudioLevel", (level: number) => {
+          setLocalAudioLevel(level)
+      })
+      .on("remoteAudioLevel", (level: number) => {
+        if (botSpeakingRef.current) {
+          setRemoteAudioLevel(level)
+        }
+      })
+      .on("userStartedSpeaking", () => {
+        // nothing to do here
+      })
+      .on("userStoppedSpeaking", () => {
+        setLocalAudioLevel(0)
```

Comment:
Just to check my understanding: the reason we explicitly set audio level to 0 here is because the Pipecat RN client might report audio levels slightly above 0 even when the user is not determined to be speaking by VAD, but we don't want those above-0 values showing up in our visualization until VAD thinks the user _is_ speaking?

Hm, wait, that can't be the reason...because "userStoppedSpeaking" is a one-time event. If we wanted the visualization to be zero-ed out until "userStartedSpeaking" we would need to track the state `isUserSpeaking` and make sure that `localAudioLevel` is 0 until `isUserSpeaking` is `true`...

So I'm not sure what the reason is.

Reply:
> Just to check my understanding: the reason we explicitly set audio level to 0 here is because the Pipecat RN client might report audio levels slightly above 0 even when the user is not determined to be speaking by VAD, but we don't want those above-0 values showing up in our visualization until VAD thinks the user is speaking?

Actually, I implemented this for the bot. So, when the bot is not speaking, the interface should behave as expected. Otherwise, you're right, we keep receiving audio levels slightly above 0.

But we don't need to do that for the local audio level. I'm going to remove it.

Reply:
done.